backend/packages/Upgrade/src/api/services/SegmentService.ts

```diff
@@ -125,32 +143,45 @@ export class SegmentService {
         this.getExperimentSegmentExclusionData(),
       ]);
 
-      const segmentsUsedList = [];
+      const segmentMap = new Map<string, string[]>();
+      segmentsData.forEach((segment) => {
+        if (segment.id === globalExcludeSegment.id) {
+          segmentMap.set(segment.id, []);
+        } else {
+          segmentMap.set(
+            segment.id,
+            segment.subSegments.map((subSegment) => subSegment.id)
+          );
+        }
+      });
+
+      const segmentsUsedList = new Set<string>();
+
+      const collectSegmentIds = (segmentId: string) => {
+        if (segmentsUsedList.has(segmentId)) return;
+        segmentsUsedList.add(segmentId);
+        const subSegmentIds = segmentMap.get(segmentId) || [];
+        subSegmentIds.forEach((subSegmentId) => collectSegmentIds(subSegmentId));
+      };
 
       if (allExperimentSegmentsInclusion) {
         allExperimentSegmentsInclusion.forEach((ele) => {
-          const subSegments = ele.segment.subSegments;
-          segmentsUsedList.push(...subSegments.map((subSegment) => subSegment.id));
+          collectSegmentIds(ele.segment.id);
+          ele.segment.subSegments.forEach((subSegment) => collectSegmentIds(subSegment.id));
```

Comment:
We don't need this line of code. The above collectSegmentIds would have called its subSegments.

Reply:
we need this line as ele.segment.id is a private segment.